test/e2e/denylist_test.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func testDenylist(t *testing.T) {
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup(t)
+
+	operatorNamespace := ctx.CreateNamespace(t, framework.KubeClient)
+	allowedNamespaces := ctx.CreateNamespace(t, framework.KubeClient)
+	deniedNamespaces := ctx.CreateNamespace(t, framework.KubeClient)
+
+	ctx.SetupPrometheusRBAC(t, operatorNamespace, framework.KubeClient)
+
+	_, err := framework.CreatePrometheusOperator(operatorNamespace, *opImage, nil, []string{deniedNamespaces}, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx.SetupPrometheusRBAC(t, deniedNamespaces, framework.KubeClient)
+	deniedPrometheus := framework.MakeBasicPrometheus(deniedNamespaces, "denied", "test", 1)
+	_, err = framework.MonClientV1.Prometheuses(deniedNamespaces).Create(deniedPrometheus)
+	if err != nil {
+		t.Fatalf("creating %v Prometheus instances failed (%v): %v", deniedPrometheus.Spec.Replicas, deniedPrometheus.Name, err)
+	}
+
+	ctx.SetupPrometheusRBAC(t, allowedNamespaces, framework.KubeClient)
+	allowed := framework.MakeBasicPrometheus(allowedNamespaces, "allowed", "test", 1)
+	allowed, err = framework.CreatePrometheusAndWaitUntilReady(allowedNamespaces, allowed)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// this is not ideal, as we cannot really find out if prometheus operator did not reconcile the denied prometheus.
+	// nevertheless it is very likely that it reconciled it as the allowed prometheus is up.
+	sts, err := framework.KubeClient.AppsV1().StatefulSets(deniedNamespaces).Get("prometheus-"+deniedPrometheus.Name, metav1.GetOptions{})
```

Comment:
As discussed: maybe we can create a ServiceMonitor in the denined ns and make sure it's not picked up by prom-operator. Just a suggestion, not sure if its the same thing?

Reply:
Very good suggestion :+1: